### DISCUSSION TITLE
allow to pull data outside dvc project

### DISCUSSION THREAD

**Discussion by link89** [treeverse/dvc#8460] [New Features & Ideas]
I am on a task on setup a CI job for a tensorflow task that need to start a tensorflow job on a HPC cluster from the CI node. In order to do this I need to send data to the HPC control node first.
In order to reduce uneccessary IO and network usage, I hope that I can just send the *.dvc file to the remote node first, and then run `dvc pull` command to fetch data to the remote node directly, instead of pulling data on the CI node and scp to the remote node.
Currently I have to run `git init` and `dvc init` before I can run `dvc pull`, it would be greate if dvc could provide a sub command that allow me to do it directly.

**Comment by dberenbaum** [comment]
Have you taken a look at [`dvc get`](https://dvc.org/doc/command-reference/get) as a way to get the data onto the remote node?

**Reply by link89** [reply]
Yeah, but it looks like dvc get don't allow to override the remote configuration.

**Reply by pared** [reply]
@link89 
- how do you want to override remote configuration? is it about credentials?
- What is a node? Is it some kind of machine that has, for example, git installed?
- is it possible to make the node perform some operations? Like running a bash script?

**Reply by link89** [reply]
Here is my use case: I am working on a solution to automate jobs running on HPC cluster, and I want to use dvc to help us to manage data.
During the development, the data are pushed to the LustreFs of the HPC culster via ssh protocol.
And when I need to run job on HPC cluster, I don't need to pull data from via ssh again, I just need to fetch data from Lustres to alluxio fuse fs directly and don't need to use the ssh protocol. I am not sure if it is a good idea to use dvc this way, but I think that would be more efficent than pulling data via ssh as I don't need to inject ssh credential for the slurm batch script.

**Reply by pared** [reply]
So in the scenario proposed by @dberenbaum you cannot `dvc get` because your "new" remote would be on this LustreFs?
And you cannot just add it to repo remotes because I assume it is not a constant existing machine but it gets created when you want to execute the job?

**Reply by link89** [reply]
I don't think [dvc get](https://dvc.org/doc/command-reference/get) have an option to select remote. And besides `dvc get` need to access git url while in most HPC cluster internet access is blocked. 
But I think `dvc get` is a good place to implement such feature by 
1. Support a new option named `--remote-url` to allow user to specific remote address on demand
2. The url field support not only git url but also local file system path.

If those 2 features could be add to `dvc` , then I can just use the following command to fetch data
```sh
dvc get --remote-url /lustre/dvc-object-store /path/to/dvc-files  /path/to/data/download/folder
```

**Reply by dberenbaum** [reply]
Would it be enough to get the URL first using [`dvc get --show-url`](https://dvc.org/doc/command-reference/get#example-getting-the-storage-url-of-a-dvc-tracked-file) and send that URL to the remote nodes? You can then download the data on there however you want.

**Reply by link89** [reply]
It won't work if you don't have access to git, most HPC cluster limits network access so that I cannot just use `dvc get`, even if you have network access, the url you get will be ssh format instead of file path as `dvc get` don't allow you to change remote on demand.

This is what I want to do:

1. `dvc push` to upload object to HPC cluster ssh://user@hpc-node/lustre/dev-object-store during development.
2. `scp -r ./data/**/*.dvc  user@hpc-node/workspace` to transfer the *.dvc file to remote machine
3. What missing here is to allow me to fetch the data (and also keep the file structure) from /lustre/dev-object-store. It would be great if dvc can provide a feature to support this.

I think it's reasonable for `dvc get` to support `--remote-url` to change remote url on demand, and support local path instead of just remote git/dvc repo. I would like to implement this if you are agree with this idea.

**Reply by link89** [reply]
I found `dvc get` already support file path instead of git url for edge case, so now what it miss is to allow to change remote url on demand
> Notice that the url provided to [dvc get](https://dvc.org/doc/command-reference/get) above is .. [dvc get](https://dvc.org/doc/command-reference/get) accepts file system paths as the "URL" to the source repo, for edge cases.

**Reply by pared** [reply]
@link89 So if you are already using `dvc` files on the lustre fs, why not checkout the data there and use `scp` to copy the data itself, not the associated `.dvc` file to worker node?

**Reply by link89** [reply]
The dvc files are on my computer and git repo, the actual data is on lustre-fs, and worker node can access the lustre-fs directly. The reason we push data to lustres-fs via ssh instead of scp of orignal files is we are using dvc to share data and manage data's version. There are other developers works on the same project, by doing this way they can easily pull data from lustre-fs to their own machines. 

When we submit job to run formal task on worker node, the woker node don't need to pull data via ssh because the it can access lustre-fs directly.  But currently there is no clear way to do this.

**Reply by pared** [reply]
Ok, I think I got it now. One more question:
worker nodes don't have `git` and `dvc` installed?


Another thing is whether it is possible to do this currently, and I believe it is, with some coding:

Have you considered using `dvc.api`?
1. In the control node - use `dvc.api.get_url` to obtain remote address of required asset - you can provide `remote` and `rev` there
2. write a script that uses `dvc.api.open` to read the assets and write them to particular worker - one case to handle is parsing directory and iteratively obtain and copy included files - `get_url` will return an address to a json file containing "relpath" relative to directory and its `md5` - here you could use `relpath` to build in-repo path and obtain the address with `get_url`.

**Reply by link89** [reply]
Yes, we have git and dvc install. But on HPC we cannot just `git clone` something as the network is blocked. I have to scp the repo to worker node. Actually we have a work around by adding a remote temporary and use dvc pull to fetch data. But I think it would be clear to have a dedicated command.
I think that's an option to use `dvc.api` , I will consider this. But I guess it is easier to do this in `dvc get` command. It's reasonable to have a way to fetch data by providing remote-url and *.dvc files.  It's like an enhanced version of `dvc get-url` that not only fetch data from remote but also reconstruct file structure according to folder that contains *.dvc files.
